consensus/src/marshal/actor.rs

```diff
@@ -577,16 +588,47 @@ impl<B: Block, E: Rng + Spawner + Metrics + Clock + GClock + Storage, V: Variant
                                     let _ = response.send(true);
                                 },
                                 Request::Finalized { height } => {
-                                    // Parse finalization
-                                    let Ok((finalization, block)) = <(Finalization<V, B::Commitment>, B)>::decode_cfg(value, &((), self.codec_config.clone())) else {
+                                    let mut reader = value;
+
+                                    // Decode the proposal to learn the epoch.
+                                    let Ok(proposal) = Proposal::<B::Commitment>::read(&mut reader) else {
+                                        let _ = response.send(false);
+                                        continue;
+                                    };
+
+                                    // Get signing scheme for epoch
+                                    let Some(signing) = self.cache.get_signing_scheme(proposal.epoch()) else {
```

Comment:
I think the way you have this code set up makes sense. I think it would be more "canonical" to instead just do a single `<(Finalization::<S, _>, B)>::decode_cfg(&mut reader, &(...))` call. The call to `finalization.verify(...)` uses `signing` which is based on the epoch of the proposal.

Generally, the codec config can just set some sane limits on how much data can be decoded. If the data is bad for whatever reason, we can [block the peer](https://github.com/commonwarexyz/monorepo/blob/3b8b06dde1c7cc4aad80661e3ead0e8afd9a244a/consensus/src/threshold_simplex/actors/batcher/actor.rs#L125) to limit how many times they can grief us in this way

Reply:
I added a comment on this.